doc/read-the-docs-site/adr/0006-term-with-type-information.md

```diff
@@ -0,0 +1,96 @@
+
+## Decision
+
+### Adding a term type with type information
+
+Currently, we have one type of terms for each PIR, PLC, and UPLC,
+and the terms don't carry type information from the type checker.
+E.g., our PIR [Term](https://github.com/input-output-hk/plutus/blob/06c5831ea5934beef2f4c4178a8c53771381f1bf/plutus-core/plutus-ir/src/PlutusIR/Core/Type.hs#L117) is:
+
+```haskell
+-- current PIR terms
+data Term tyname name uni fun a =
+                        -- Plutus Core (ish) forms, see note [Declarations in Plutus Core]
+                          Let a Recursivity (NonEmpty (Binding tyname name uni fun a)) (Term tyname name uni fun a)
+                        | Var a name
+                        | TyAbs a tyname (Kind a) (Term tyname name uni fun a)
+                        | LamAbs a name (Type tyname uni a) (Term tyname name uni fun a)
+                        | Apply a (Term tyname name uni fun a) (Term tyname name uni fun a)
+                        | Constant a (PLC.Some (PLC.ValueOf uni))
+                        | Builtin a fun
+                        | TyInst a (Term tyname name uni fun a) (Type tyname uni a)
+                        | Error a (Type tyname uni a)
+                        | IWrap a (Type tyname uni a) (Type tyname uni a) (Term tyname name uni fun a)
+                        | Unwrap a (Term tyname name uni fun a)
+                        deriving stock (Functor, Show, Generic)
+```
+
+We proposed to add a term type that is distinguished from the current type in that it carries [
+type](https://github.com/input-output-hk/plutus/blob/06c5831ea5934beef2f4c4178a8c53771381f1bf/plutus-core/plutus-core/src/PlutusCore/Core/Type.hs#L93)
+information.
+The proposed term types include a generic term type,
+a term type with type information, and a term type without type information (what we have currently).
+
+```haskell
+-- the new generic term: it has an additional parameter: tyinfo
+data GenericTerm tyinfo tyname name uni fun a =
+    Let tyinfo a Recursivity (NonEmpty (Binding tyname name uni fun a)) (Term tyname name uni fun a)
+    | Var tyinfo a name
+    | TyAbs tyinfo a tyname (Kind a) (Term tyname name uni fun a)
+    | LamAbs tyinfo a name (Type tyname uni a) (Term tyname name uni fun a)
+    | Apply tyinfo a (Term tyname name uni fun a) (Term tyname name uni fun a)
+    | Constant tyinfo a (PLC.Some (PLC.ValueOf uni))
+    | Builtin tyinfo a fun
+    | TyInst tyinfo a (Term tyname name uni fun a) (Type tyname uni a)
+    | Error tyinfo a (Type tyname uni a)
+    | IWrap tyinfo a (Type tyname uni a) (Type tyname uni a) (Term tyname name uni fun a)
+    | Unwrap tyinfo a (Term tyname name uni fun a)
+    deriving stock (Functor, Show, Generic)
+
+-- the new term type that is similar to our current implementation, with tyinfo set to ()
+type Term = GenericTerm () tyname name uni fun a
+
+-- the new term type that has type information
+type TermWithTypeInfo = GenericTerm (Type tyname uni ann) tyname name uni fun a
```

Comment:
Here we do need the type parameters, but they need to be bound on the LHS. Plus there's that confusion of `ann` vs `a`. I'm much on the side of always using `ann` instead of `a` to convey the semantics.

Reply:
Totally agree about `ann`!